library/core/src/option.rs

```diff
@@ -2545,3 +2546,41 @@ impl<T> Option<Option<T>> {
         }
     }
 }
+
+impl<T, const N: usize> [Option<T>; N] {
+    /// Transposes a `[Option<T>; N]` into a `Option<[T; N]>`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(option_array_transpose)]
+    /// # use std::option::Option;
+    ///
+    /// let data = [Some(0); 1000];
+    /// let data: Option<[u8; 1000]> = data.transpose();
+    /// assert_eq!(data, Some([0; 1000]));
+    ///
+    /// let data = [Some(0), None];
+    /// let data: Option<[u8; 2]> = data.transpose();
+    /// assert_eq!(data, None);
+    /// ```
+    #[inline]
+    #[unstable(feature = "option_array_transpose", issue = "130828")]
+    pub fn transpose(self) -> Option<[T; N]> {
+        let mut array: MaybeUninit<[T; N]> = MaybeUninit::uninit();
+        let array_mut = array.as_mut_ptr();
+
+        for (i, e) in self.into_iter().enumerate() {
+            match e {
+                // SAFETY: `i` is always inbounds
+                Some(e) => unsafe {
+                    *(*array_mut).get_unchecked_mut(i) = e;
```

Comment:
This can be avoided by collecting into `[MaybeUninit<T>; N]` instead, and using `.zip(array.iter_mut())`. Then, you can use the unstable `MaybeUninit::array_assume_init`.

Reply:
This also introduces a `&mut uninit`. Not sure if that is UB or not but I'd probably go the `[MaybeUninit<T>; N]` route since that's one less unsafe block and optimizes equally: https://godbolt.org/z/d8d7a64h5.

Reply:
I'm talking about https://godbolt.org/z/G4oYxEMMY, no need to produce a `&mut uninit`, only a `&mut MaybeUninit<T>` which is overwritten by a `MaybeUninit::new(val)`.

Reply:
@GnomedDev I was actually referring to the PR code 😅. I think we are in an agreement that using the array maybeuninit api would be more ideal here :)

Reply:
Looking at both versions with the `String` type, [godbolt](https://godbolt.org/z/P6eeeK9rv), your version generated ~2x more assembly than mine[^1], I don't know why that would be the case. I also don't know if it's better for perf or not.

[^1]: Your version seems to be consistently bigger, not always by ~2x, sometimes only 3/4 ins

Reply:
> Looking at both versions with the `String` type, [godbolt](https://godbolt.org/z/P6eeeK9rv), your version generated ~2x more assembly than mine[1](#user-content-fn-1-209ddd200ca274b23436615f2c808656), I don't know why that would be the case. I also don't know if it's better for perf or not.
> 
> ## Footnotes
> 1. Your version seems to be consistently bigger, not always by ~2x, sometimes only 3/4 ins [↩](#user-content-fnref-1-209ddd200ca274b23436615f2c808656)

In the link you have provided, the PR version did not get vectortized, while the `.zip()` one did. This probably correlates with the code size difference.

May I propose yet another version:

```rust
fn transpose3<T, const N: usize>(val: [Option<T>; N]) -> Option<[T; N]> {
    val.try_map(core::convert::identity)
}
```

https://godbolt.org/z/445sPh88v

Reply:
I've done some benchmarks using criterion between the three versions, [gist of code + raw results](https://gist.github.com/Urgau/98edfc2b86f0ea72b0639dee1bd34ae0), and it seems like the last version you proposed is slightly faster overall, while also be safer, I will go with it.

![image](https://github.com/user-attachments/assets/fd037018-e4e8-4561-92f5-9799e1d477a7)